In [1]:
import numpy as np
import pandas as pd
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.proportion import proportions_ztest, proportion_effectsize
from statsmodels.stats.power import NormalIndPower

np.random.seed(42)

In [ ]:
counts = np.array([120,85,110,85])
expected_counts = np.array([100,100,100,100])
alpha = 0.05

In [36]:

chi2_stat, p_value = st.chisquare(counts, f_exp=expected_counts)
ddof = len(counts) - 1
print(f"Chi-square statistic: {chi2_stat:.4f}")
print(f"P-value: {p_value:.4f}")
print(f"Degrees of freedom: {ddof}")
print("Reject the null hypothesis" if p_value < alpha else "Fail to reject the null hypothesis")

Chi-square statistic: 9.5000
P-value: 0.0233
Degrees of freedom: 3
Reject the null hypothesis


In [ ]:
import pandas as pd

contingency_table = pd.DataFrame({
    "Free": [90, 110],
    "Basic": [60, 140],
    "Premium": [30, 170]
}, index=["Churned", "Retained"])

print(contingency_table)

          Free  Basic  Premium
Churned     90     60       30
Retained   110    140      170


In [40]:
chi2, p_value, dof, expected = st.chi2_contingency(contingency_table)
print(f"Chi-square statistic: {chi2:.4f}")
print(f"P-value: {p_value:.4f}")
print(f"Degrees of freedom: {dof}")
print(f"Expected counts:\n{pd.DataFrame(expected, index=contingency_table.index, columns=contingency_table.columns)}")
print("Reject the null hypothesis" if p_value < alpha else "Fail to reject the null hypothesis")

Chi-square statistic: 42.8571
P-value: 0.0000
Degrees of freedom: 2
Expected counts:
           Free  Basic  Premium
Churned    60.0   60.0     60.0
Retained  140.0  140.0    140.0
Reject the null hypothesis


from the results, as Chi-square is 42.8571 there's and p-value is so small, our hypothesis is rejected and that means the type of the subscription affects whether users churn or retain. Better the subscription lowerthe churn and more the retain

In [29]:
def cramers_v(chi2, n, min_dim):
    """
    Returns Cramér's V.

    Parameters
    ----------
    chi2 : float
        Chi-square statistic.
    n : int
        Total number of observations.
    min_dim : int
        min(rows, cols) - 1 of the contingency table.
    """
    return np.sqrt(chi2 / (n * min_dim))

In [ ]:
n = contingency_table.values.sum()
min_dim = min(contingency_table.shape) - 1
cramers_v_value = cramers_v(chi2, n, min_dim)   
print(f"Cramér's V: {cramers_v_value:.4f}")
print("Effect size interpretation:", "Small effect" if cramers_v_value < 0.1 else "Medium effect" if cramers_v_value < 0.3 else "Large effect")n = contingency_table.values.sum()
min_dim = min(contingency_table.shape) - 1
cramers_v_value = cramers_v(chi2, n, min_dim)   
print(f"Cramér's V: {cramers_v_value:.4f}")
print("Effect size interpretation:", "Small effect" if cramers_v_value < 0.1 else "Medium effect" if cramers_v_value < 0.3 else "Large effect")

Cramér's V: 0.2673
Effect size interpretation: Medium effect


In chi-square we find out if theres a relationship between them, but with help of Cramer's V we can get how much of a relationship there is. Since Cramer's V is equal to 0.2673 it means there's a medium relationship